umbraco/Umbraco-CMS
# Fix Block List settings exception and optimize PVCs

### Prerequisites


If there's an existing issue for this PR then this fixes: https://github.com/umbraco/Umbraco-CMS/issues/11712.

### Description
Since PR https://github.com/umbraco/Umbraco-CMS/pull/8854 (v8.17+ and v9), getting the correct settings model type relied on having generated models available: if this wasn't the case, an `InvalidOperationException` was thrown.

The root cause of this seems to be the use of [`IPublishedModelFactory.MapModelType()`](https://github.com/umbraco/Umbraco-CMS/blob/bf41c2eeaaf7137ab8223b6557514e5ca84abd5b/src/Umbraco.Core/Models/PublishedContent/PublishedModelFactory.cs#L125-L126). Unfortunately the `IPublishedModelFactory` interface doesn't expose another way to directly get the model type based on a content type alias (without throwing this exception), but I've found an indirect way by using `CreateModelList()` and getting the generic type argument of the list 🙌🏻 

Because this involves a but more work to get the model type and there was also a comment that we should optimize/cache the creation of new `BlockListItem<,>` types, this PR makes sure that is now taken care of.

I also noticed Nested Content used `ModelType.For()` in `GetPropertyValueType()`, but that only seems to be used when generating the models (and therefore will not throw the exception). There were some other things that could be improved though, which I've included in this PR as well.

Testing steps:
- Create a document type with 2 properties:
  - `blocks` using the Block List editor with 2 configured blocks:
    1. Content model: `Text`, settings model: `TextSettings`
    2. Content model: `Text2`
  - `nested` using Nested Content with 2 element types:
    1. `Text`
    2. `Text2`
- Add the following to the template of that document type:

```cshtml
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage
@using Umbraco.Cms.Web.Common.PublishedModels;
@using Umbraco.Cms.Core.Models.Blocks;
@{
	Layout = null;
}
<h1>Blocks: @Model.Value("blocks").GetType()</h1>
@foreach (var blockListItem in Model.Value<BlockListModel>("blocks"))
{
    <h2>Block: @blockListItem.GetType()</h2>
    <h3>Block content: @blockListItem.Content.GetType()</h3>
    @if (blockListItem.Settings != null)
    {
       <h3>Block settings: @blockListItem.Settings.GetType()</h3>
    }
}
<h1>Nested: @Model.Value("nested").GetType()</h1>
@foreach (var nestedContentItem in Model.Value<IEnumerable<IPublishedElement>>("nested"))
{
    <h2>Content: @nestedContentItem.GetType()</h2>
}
```

After adding a `Text` and `Text2` content item to both editors and publishing the page, you should get the following output when using the default `Umbraco:CMS:ModelsBuilder:ModelsMode` that's set to `InMemoryAuto`:
```
Blocks: Umbraco.Cms.Core.Models.Blocks.BlockListModel
Block: Umbraco.Cms.Core.Models.Blocks.BlockListItem`2[Umbraco.Cms.Web.Common.PublishedModels.Text,Umbraco.Cms.Web.Common.PublishedModels.TextSettings]
Block content: Umbraco.Cms.Web.Common.PublishedModels.Text
Block settings: Umbraco.Cms.Web.Common.PublishedModels.TextSettings
Block: Umbraco.Cms.Core.Models.Blocks.BlockListItem`2[Umbraco.Cms.Web.Common.PublishedModels.Text2,Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement]
Block content: Umbraco.Cms.Web.Common.PublishedModels.Text2

Nested: System.Collections.Generic.List`1[Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement]
Content: Umbraco.Cms.Web.Common.PublishedModels.Text
Content: Umbraco.Cms.Web.Common.PublishedModels.Text2
```

And changing the ModelsBuilder mode to `Nothing` and restarting should now show:
```
Blocks: Umbraco.Cms.Core.Models.Blocks.BlockListModel
Block: Umbraco.Cms.Core.Models.Blocks.BlockListItem`2[Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement,Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement]
Block content: Umbraco.Cms.Core.PublishedCache.PublishedElement
Block settings: Umbraco.Cms.Core.PublishedCache.PublishedElement
Block: Umbraco.Cms.Core.Models.Blocks.BlockListItem`2[Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement,Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement]
Block content: Umbraco.Cms.Core.PublishedCache.PublishedElement

Nested: System.Collections.Generic.List`1[Umbraco.Cms.Core.Models.PublishedContent.IPublishedElement]
Content: Umbraco.Cms.Core.PublishedCache.PublishedElement
Content: Umbraco.Cms.Core.PublishedCache.PublishedElement
```
